sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using Azure.Core.Json;
+
+namespace Azure.Core.Dynamic
+{
+    public partial class DynamicData
+    {
+        internal class AllowList
+        {
+            public static void AssertAllowedType<T>(T value)
+            {
+                if (!IsAllowedType(typeof(T)))
+                {
+                    throw new NotSupportedException($"Type is not currently supported: '{typeof(T)}'.");
+                }
+            }
+
+            public static bool IsAllowedType(Type type)
+            {
+                if (IsAllowedPrimitive(type))
+                {
+                    return true;
+                }
+
+                if (IsAllowedCollectionType(type))
+                {
+                    return true;
+                }
+
+                return IsAllowedAnonymousType(type, new HashSet<Type>());
+            }
+
+            // Primitive means "not a collection and not a POCO"
+            private static bool IsAllowedPrimitive(Type type)
+            {
+                return
+                    type.IsPrimitive ||
+                    type == typeof(decimal) ||
+                    type == typeof(string) ||
+                    type == typeof(DateTime) ||
+                    type == typeof(DateTimeOffset) ||
+                    type == typeof(TimeSpan) ||
+                    type == typeof(Uri) ||
+                    type == typeof(Guid) ||
+                    type == typeof(ETag) ||
+
+                    // Allowable for DynamicData assignments.
+                    type == typeof(JsonElement) ||
+                    type == typeof(JsonDocument) ||
+                    type == typeof(MutableJsonDocument) ||
+                    type == typeof(MutableJsonElement) ||
+                    type == typeof(DynamicData) ||
+
+                    // We allow object so we can allow heterogenous object[] and
+                    // Dictionary<string, object>. It also means we have to validate
+                    // the types of values as well.
+                    type == typeof(object);
```

Comment:
Can we remove this and special case for heterogenous collections? It seems like this will let an object instance value pass, and we don't want that,

Reply:
Done.